uniffi_core/src/ffi/rustfuture.rs

```diff
@@ -0,0 +1,642 @@
+//! known.
+//!
+//! # The big picture
+//!
+//! This section will explain how the entire workflow works.
+//!
+//! Let's consider the following Rust function:
+//!
+//! ```rust,ignore
+//! #[uniffi::export]
+//! async fn hello() -> bool {
+//!     true
+//! }
+//! ```
+//!
+//! In Rust, this `async fn` syntax is strictly equivalent to:
+//!
+//! ```rust,ignore
+//! #[uniffi::export]
+//! fn hello() -> impl Future<Output = bool> { /* … */ }
+//! ```
+//!
+//! Once this is understood, it becomes obvious that an `async` function
+//! returns a `Future`.
+//!
+//! This function will not be modified, but new functions with a C ABI will be
+//! created, as such:
+//!
+//! ```rust,ignore
+//! /// The `hello` function, as seen from the outside. It returns a `Future`, or
+//! /// more precisely, a `RustFuture` that wraps the returned future.
+//! #[no_mangle]
+//! pub extern "C" fn _uniffi_hello(
+//!     call_status: &mut ::uniffi::RustCallStatus
+//! ) -> Option<Box<::uniffi::RustFuture<bool, ::std::convert::Infallible>>> {
+//!     ::uniffi::call_with_output(call_status, || {
+//!         Some(Box::new(::uniffi::RustFuture::new(async move {
+//!             Ok(hello().await)
+//!         })))
+//!     })
+//! }
+//! ```
+//!
+//! The second generated function is the _poll function_:
+//!
+//! ```rust,ignore
+//! /// The function to poll the `RustFuture` returned by `_uniffi_hello`.
+//! #[no_mangle]
+//! pub extern "C" fn _uniffi_hello_poll(
+//!     future: Option<&mut ::uniffi::RustFuture<bool, ::std::convert::Infallible>>,
+//!     waker: Option<NonNull<::uniffi::RustFutureForeignWakerFunction>>,
+//!     waker_environment: *const ::uniffi::RustFutureForeignWakerEnvironment,
+//!     polled_result: &mut <bool as ::uniffi::FfiReturn>::FfiType,
+//!     call_status:: &mut ::uniffi::RustCallStatus,
+//! ) -> bool {
+//!     ::uniffi::ffi::uniffi_rustfuture_poll(future, waker, waker_environment, polled_result, call_status)
+//! }
+//! ```
+//!
+//! Let's analyse this function because it's an important one:
+//!
+//! * First off, this _poll FFI function_ forwards everything to
+//!   [`uniffi_rustfuture_poll`]. The latter is generic, while the former has been
+//!   monomorphised by the procedural macro.
+//!
+//! * Second, it receives the `RustFuture` as an `Option<&mut RustFuture<_>>`. It
+//!   doesn't take ownership of the `RustFuture`! It borrows it (mutably). It's
+//!   wrapped inside an `Option` to check whether it's a null pointer or not.
+//!
+//! * Third, it receives a _waker_ as a pair of a _function pointer_ plus its
+//!   _environment_, if any; a null pointer is purposely allowed for the environment.
+//!   This waker function lives on the foreign language side. We will come back
+//!   to it in a second.
+//!
+//! * Fourth, it receives an in-out `polled_result` argument, that is filled with the
+//!   polled result if the future is ready.
+//!
+//! * Finally, the classical `call_status`, which is part of the calling API of `uniffi`.
```

Comment:
- Can you explain how this interacts with the return value more?  Is this also only set if the future is ready, like `polled_result`?
- Please document the return value.  I assume true means ready and false means not ready.
- I prefer having a `u8` as the return value and defining constants for Ready/Pending.  That way it's very clear and explicit what the ABI representation is.

Reply:
Doc' updated. I don't mark this as resolved because I need to add constant for `bool/u8`.